Open PR against gh-merge-base

heaths opened the pull request:
Partly resolves issue #8979 by checking for a `gh-merge-base` branch tag and using that as though it were passed to `gh pr create --base`.

heaths committed:
Open PR against gh-merge-base

Partly resolves issue #8979 by checking for a `gh-merge-base` branch tag and using that as though it were passed to `gh pr create --base`.

heaths commented:
I tested this against a repo I created with divergent `main` and `feature/feat1` branches, and a `task1` branch created from the latter with a manually set `gh-merge-base`: https://github.com/heaths/test-repo/pull/2

That said, we should probably wait to hear from @alexr00 or whoever manages the VSCode GitHub PR & Issues extension to possibly coordinate efforts. Or we could proceed and use those as fallbacks, but it seems a bit silly for the GitHub extension and the CLI to diverge.

heaths committed:
Set gh-merge-base from `issue develop`

williammartin commented:
Hey @heaths, just filling in the communication gap here. This is on my radar but we're pretty busy in the lead up to universe next week. I expect it will be prioritised right after that.

williammartin commented on git/client.go:
```diff
@@ -19,6 +19,9 @@ import (
 )
 
+// MergeBaseConfig is the configuration setting to keep track of the PR target branch.
+const MergeBaseConfig = "gh-merge-base"
```
**Question**

Is `merge` the best word for this name? Does the vscode extension use it for other things than the PR base? Can you think of a reason we would want to do something else in the future? If not, maybe `gh-pr-base` is clearer?

heaths replied:
It's about consistency not only with other tooling - both VSCode (the legacy name) and the GitHub Issues and Pull Requests extension - but also with the CLI's only wording: from `gh pr create --help`:

```text
  -B, --base branch          The branch into which you want your code merged
```

Ultimately, it's what branch your topic branch gets merged into. Not married to it, but I also see no strong reason to change it. I appreciate the cost of coordination may not be worth it, but I think sharing common wording helps understanding.

williammartin replied:
I'm happy with this reasoning.

williammartin commented on pkg/cmd/issue/develop/develop.go:
```diff
@@ -171,6 +171,14 @@ func developRunCreate(opts *DevelopOptions, apiClient *api.Client, issueRepo ghr
 
+	// Remember which branch to target when creating a PR.
+	if opts.BaseBranch != "" {
+		err = opts.GitClient.SetBranchConfig(ctx.Background(), branchName, git.MergeBaseConfig, opts.BaseBranch)
```
**Concern**

I'm a bit nervous about this purely because it's a breaking change.

Before this, development branches would always go back to the default branch unless otherwise specified with a flag. I don't have a clear answer for whether there are cases out there where people expect to create a development branch from a different branch than they want to PR back into.

Maybe we just bite the bullet and if someone complains, we can apologise. This definitely seems like the more natural default.

What do you think?

heaths replied:
I appreciate it's a change, but seems it's a change for good. The CLI created the topic branch from the `opts.BaseBranch` in the couple of code blocks above. If it were to target the default branch, the commit graph could be a real mess: the base branch may not be merged or, worse, has diverged from its base branch entirely e.g., a hotfix or feature branch, or just a branch for stabilization aka "release branch".

But, yes, it is a change. What if, as compensation for trying to do the, IMO, right thing here, interactive prompts to create a PR ask for the base (merge) branch with the current (or default) as the default value if they just hit Enter?

williammartin replied:
I don't think that we would target the default branch explicitly, just not set `gh-merge-base` at all, leaving the behaviour as-is. That said, I'm inclined to bite the bullet on this one. It makes sense to me that the default is to target the branch that we used as a base.

If we hear complaints then we can change it back and perhaps introduce a config setting for permanently chjoosing what to do.

> What if, as compensation for trying to do the, IMO, right thing here, interactive prompts to create a PR ask for the base (merge) branch with the current (or default) as the default value if they just hit Enter?

I think that makes sense to me. Mostly this is going to be an issue in the interactive case because it's part of `gh issue develop` which is almost always going to involve a human at the keyboard.

williammartin reviewed (commented):
@heaths

Generally, I think this is very close. One question and one concern I'd like your input on.

Since this had been open for so long, and I wanted to make use of some testing utilities in `trunk`, I created https://github.com/cli/cli/pull/10004 since I didn't want to force push over your forked branch. I added some docs and a couple of acceptance tests. We're internally discussing what we want to do with this test suite that we started building for a different purpose, but I'm finding it useful to test some of these black box behaviours, and as demonstrations. Don't feel the need to add them yourself, they are just for the maintainers right now.

I don't feel that strongly about tying ourselves to what VSCode uses right now. Primarily my concern is that our teams are pretty distant, and I'm concerned about the increased cost of coordination. I'm happy to wait until people say they really want it to figure something out (maybe never?).

williammartin commented:
If we address the `pr create` prompt, and bring in the changes from my PR, I think this is good to go. Thanks.

heaths commented:
I rebased and cherry-picked your changes, sans an errant whitespace issue and a couple of grammatical fixes. I'll try to get to the `gh pr create` interactive change yet tonight.

heaths committed:
Merge changes from #10004

Merges changes from @williammartin including acceptance tests and word changes.

Co-authored-by: William Martin <williammartin@github.com>

heaths committed:
Confirm auto-detected base branch

If interactive, confirm the automatically configured gh-merge-branch or, if not configured, the default branch.

Based on PR feedback.

heaths reviewed (commented):
@williammartin I implemented the prompt, did some manual tests, and added a prompt test. Unfortunately - though practical - the initial message about creating the PR from the topic branch to the base branch is based on the configured branch. I added a comment to that effect.

![image](https://github.com/user-attachments/assets/b099240a-1d19-45de-8d7f-0c86acc41d81)

williammartin commented on pkg/cmd/pr/create/create.go:
```diff
@@ -340,7 +342,7 @@ func createRun(opts *CreateOptions) (err error) {
 	}
 
-	if !opts.EditorMode && (opts.FillVerbose || opts.Autofill || opts.FillFirst || (opts.TitleProvided && opts.BodyProvided)) {
+	if !opts.EditorMode && (opts.FillVerbose || opts.Autofill || opts.FillFirst || (opts.TitleProvided && opts.BodyProvided && ctx.BaseBranch != "")) {
```
**Question**

Can `ctx.BaseBranch` ever be `""` under usual circumstances at this point? It seems like various things prior and after assume that it has a value.

heaths replied:
It shouldn't, at least through my review and tests.

williammartin committed:
Revert "Confirm auto-detected base branch"

This reverts commit e021a072850dac409ce011b01238cd6b3931f7ac.

williammartin reviewed (approved):
Thanks for doing the prompting on PR create. I played around with it myself and decided that:
 * I didn't like the prompt appearing under the normal flow today (i.e. when there is no `gh-merge-base` config), because it's adding an extra hop that is probably unchanged 99% of the time)
 * Didn't like that the prompting would diverge between having `gh-merge-base` configured, and not.
 * Didn't like the extra hop that would be required of everyone using `gh-merge-base`.
 * Didn't think that after maybe some initial annoyance, that it provided that much more value over the current text that mentions the branch explicitly:

```
Creating pull request for williammartin:merge-base-pr into feature in williammartin-test-org/williammartin-test-org-test-repo-rename
```
 * Didn't like that this static text could later be incorrect on scrollback because it was later changed via interactivity.
 
 So all those things considered, I decided to revert the last commit, hope that makes sense. Sorry for the extra work.

heaths commented:
I see you already merged it, but I woke this morning - after going to bed with an uneasy feeling I couldn't quite place about this - thinking that this prompt changes the flow quite a bit. It's always going to prompt unless you pass `--base`, so should we maybe instead prompt to set the `gh-merge-base` when running `gh issue develop`, or finish the implementation and decide then?

williammartin commented:
> so should we maybe instead prompt to set the gh-merge-base when running gh issue develop, or finish the implementation and decide then?

Hmm, really not sure. Prompting could be good, but I think it only makes sense if we also introduce a flag to avoid it in the non-interactive flow. I don't really like the lack of consistency having an option one way and not the other.

> or finish the implementation and decide then?

What implementation do you still have in mind? I know in the issue we talked about following the reflog algorithm, presumably on `pr create` with a new flag? How much extra value does this provide to you over the feature delivered here?

heaths commented:
> What implementation do you still have in mind? I know in the issue we talked about following the reflog algorithm, presumably on `pr create` with a new flag? How much extra value does this provide to you over the feature delivered here?

I never use the `gh issue develop`. I suspect a lot don't. I always `git checkout -b {new-branch} ({optional head e.g., FETCH_HEAD})` and then later `gh pr create`. We discussed above about the automatic resolution, which I'd like to finish.

I still think this needs to be in the GraphQL to find the closest ancestor known to GitHub in the upstream repo. What you have locally doesn't really matter. As long-time user of git, I often have various local branches I use for prototyping or if I want to try a couple different approaches so I can easily pick one.

heaths commented:
@jtmcg was surprised this wasn't mentioned in <https://github.blog/open-source/git/how-the-github-cli-can-now-enable-triangular-workflows/>, or is this too specific of a workflow i.e., targeting feature branches? I know we're definitely not unique in that, but a "hero scenario" it is not.

That said, the blog post reminded me that the complete E2E solution for #8979 hasn't been finalized. Are we still going to automatically set the `gh-merge-base` a la VSCode/GitHub Issues and Pull Request?
